presentations/vocabulary_types_2023_11_8.md

```diff
@@ -0,0 +1,241 @@
+We can combine vocabulary types to express combined idioms:
+
+| Idiom | Type |
+|---|---|
+| Variable-count, multiple, nullable instances of a type `T`| `std::vector<std::optional<T>>`|
+| Key-accessed, instances of one of a closed-set of types `Ts...`| `std::map<Key, std::variant<Ts...>>`|
+
+---
+
+# Incomplete types
+
+We may want a composite to contain an instance of an incomplete type, either directly or though the use of existing vocabulary types.
+
+Incomplete type support is needed for defining recursive data structures, supporting open-set polymorphism, and hiding implementation detail.
+
+Incomplete types are supported by storing the object of the incomplete type outside of the composite object.
+
+Stroring an object outside of a composite type can also be used to optimize use of cache (hot-cold-splitting).
+
+---
+
+## Incomplete types using pointers
+
+Since we want to store the object of incomplete type outside of our composite type,
+pointers are the first thing we may reach for:
+
+```c++
+class Composite {
+    T* ptr_;
+    // ...
+}
+```
+
+Pointers are a poor fit for owned data, we'll need to implement all special member functions and manually check const-qualified methods - const does not propagate to a pointee when the pointer is accessed through a const-access path.
+
+---
+
+## Incomplete types using `std::unique_ptr<T>`
+
+Unique pointers are a little better:
+
+```c++
+class Composite {
+    std::unique_ptr<T> ptr_;
+    // ...
+}
+```
+
+We don't need to implement move construction, move assignment or destruction. The compiler will implicitly delete the copy constructor and copy assignement operator; either the composite is non-copyable or we implement the copy constructor and assignment ourselves.
+
+Const does not propagate through `std::unique_ptr`, we must manually check const-qualified methods for correctness.
+
+---
+
+## Incomplete types using `std::shared_ptr<T>`
+
+Shared pointers don't model the right thing:
+
+```c++
+class Composite {
+    std::shared_ptr<T> ptr_;
+    // ...
+}
+```
+
+We don't need to implement any special member functions, but we have other issues.
+
+The compiler-generated copy constructor and assignment operator will give rise to multiple composite objects that share the same components: that's not ownership.
+
+Const does not propagate through `std::shared_ptr<T>`, we must manually check const-qualified methods for correctness.
+
+---
+
+## Incomplete types using `std::shared_ptr<const T>`
+
+Shared pointers to const might work:
+
+```c++
+class Composite {
+    std::shared_ptr<const T> ptr_;
+    // ...
+}
+```
+
+Again the compiler-generated copy constructor and assignment operator will give rise to multiple composite objects that share the same components, _but_ the shared components 
+are immutable. 
+
+If our hierarchy of composites is const-correct, immutability will hide the fact that data is shared.
+
+We can't call non-const qualified member function on our component `ptr_`.
```

Comment:
Tiny nitpick: You can call non-const qualified member function**s** on `ptr_`. What you can't do is call non-const qualified member functions on `*ptr_` (the pointee).

Reply:
Thanks, this is no nitpick but a significant mistake - well spotted!